### PR TITLE
Development 7.2

### PR DESCRIPTION
**Pull request by vamshin**
*Description of changes:*
- JNI memory leak fixes
- _opendistro/_knn/_stats api
-  code refactoring
- performance improvements
- use Guava cache 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -88,20 +88,6 @@ task release(type: Copy, group: 'build') {
     includeEmptyDirs = false
     // ES versions < 6.3 have a top-level elasticsearch directory inside the plugin zip which we need to remove
```
Comment (jmazanec15) [inline build.gradle]: Line unnecessary as we do not support < 6.3

File: jni/src/v1736/com_amazon_opendistroforelasticsearch_knn_index_v1736_KNNIndex.cpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
+#include "com_amazon_opendistroforelasticsearch_knn_index_v1736_KNNIndex.h"
```
Comment (jmazanec15) [inline jni/src/v1736/com_amazon_opendistroforelasticsearch_knn_index_v1736_KNNIndex.cpp:16]: Need to add license headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNCircuitBreaker.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
+package com.amazon.opendistroforelasticsearch.knn.index;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNCircuitBreaker.java:16]: Need to add license headers.

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNSettings.java [hunk 4]
```diff
@@ -0,0 +1,319 @@
+package com.amazon.opendistroforelasticsearch.knn.index;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNSettings.java:16]: License header

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/util/KNNConstants.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.amazon.opendistroforelasticsearch.knn.index.util;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/index/util/KNNConstants.java:16]: license headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/rest/RestKNNStatsHandler.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.rest;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/rest/RestKNNStatsHandler.java:16]: license headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/KNNStat.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.stats;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/KNNStat.java:16]: license headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/KNNStats.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.stats;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/KNNStats.java:16]: License headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/StatNames.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.stats;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/StatNames.java:16]: license headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/suppliers/KNNCacheSupplier.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.stats.suppliers;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/suppliers/KNNCacheSupplier.java:16]: License headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/suppliers/KNNCircuitBreakerSupplier.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.stats.suppliers;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/suppliers/KNNCircuitBreakerSupplier.java:16]: License headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/suppliers/KNNInnerCacheStatsSupplier.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.stats.suppliers;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/suppliers/KNNInnerCacheStatsSupplier.java:16]: License headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsAction.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.transport;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsAction.java:16]: License headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsNodeRequest.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.transport;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsNodeRequest.java:16]: License headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsNodeResponse.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.transport;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsNodeResponse.java:16]: License headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsRequest.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.transport;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsRequest.java:16]: License headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsResponse.java [hunk 17]
```diff
@@ -0,0 +1,86 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.transport;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsResponse.java:16]: License headers

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsTransportAction.java [hunk 18]
```diff
@@ -0,0 +1,96 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.transport;
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsTransportAction.java:16]: License headers

File: src/test/java/com/amazon/opendistroforelasticsearch/knn/index/BaseKNNIntegTestIT.java [hunk 19]
```diff
@@ -0,0 +1,81 @@
+package com.amazon.opendistroforelasticsearch.knn.index;
```
Comment (jmazanec15) [inline src/test/java/com/amazon/opendistroforelasticsearch/knn/index/BaseKNNIntegTestIT.java:16]: License header

File: src/test/java/com/amazon/opendistroforelasticsearch/knn/index/KNNESSettingsTestIT.java [hunk 20]
```diff
@@ -0,0 +1,135 @@
+package com.amazon.opendistroforelasticsearch.knn.index;
```
Comment (jmazanec15) [inline src/test/java/com/amazon/opendistroforelasticsearch/knn/index/KNNESSettingsTestIT.java:16]: License headers

**Review by jmazanec15 — COMMENTED** [review]
Left a comment on all files missing license headers. Outside of that LGTM. Will create an issue to add a check in gradle build for license headers.

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNCircuitBreaker.java (hunk 3 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNCircuitBreaker.java:16]: Done

File: build.gradle (hunk 1 above)
Comment (vamshin) [reply build.gradle]: Done

File: jni/src/v1736/com_amazon_opendistroforelasticsearch_knn_index_v1736_KNNIndex.cpp (hunk 2 above)
Comment (vamshin) [reply jni/src/v1736/com_amazon_opendistroforelasticsearch_knn_index_v1736_KNNIndex.cpp]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/util/KNNConstants.java (hunk 5 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/index/util/KNNConstants.java:16]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNCircuitBreaker.java (hunk 3 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNCircuitBreaker.java]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNSettings.java (hunk 4 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNSettings.java:16]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsRequest.java (hunk 16 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsRequest.java:16]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsResponse.java (hunk 17 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsResponse.java:16]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsTransportAction.java (hunk 18 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/transport/KNNStatsTransportAction.java:16]: Done

File: src/test/java/com/amazon/opendistroforelasticsearch/knn/index/BaseKNNIntegTestIT.java (hunk 19 above)
Comment (vamshin) [reply src/test/java/com/amazon/opendistroforelasticsearch/knn/index/BaseKNNIntegTestIT.java:16]: Done

File: src/test/java/com/amazon/opendistroforelasticsearch/knn/index/KNNESSettingsTestIT.java (hunk 20 above)
Comment (vamshin) [reply src/test/java/com/amazon/opendistroforelasticsearch/knn/index/KNNESSettingsTestIT.java:16]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/rest/RestKNNStatsHandler.java (hunk 6 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/rest/RestKNNStatsHandler.java:16]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/KNNStat.java (hunk 7 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/KNNStat.java:16]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/KNNStats.java (hunk 8 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/KNNStats.java:16]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/StatNames.java (hunk 9 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/StatNames.java:16]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/suppliers/KNNCircuitBreakerSupplier.java (hunk 11 above)
Comment (vamshin) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/stats/suppliers/KNNCircuitBreakerSupplier.java:16]: Done

**Review by jmazanec15 — APPROVED** [review]

**Comment by vamshin** [comment]
Incorporated comments. Merged manually

**Pull request closed without merging**